// @vouched/client/dist/common/errors.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var UNKNOWN_ERROR_CODE = exports.UNKNOWN_ERROR_CODE = 'UNKNOWN_ERROR';
var SYSTEM_ERROR_MESSAGE = exports.SYSTEM_ERROR_MESSAGE = 'Oops, we encountered a problem';
var BAD_USER_INPUT = exports.BAD_USER_INPUT = 'BAD_USER_INPUT';

var AuthenticationErrorType = exports.AuthenticationErrorType = 'AuthenticationError';
var ConnectionErrorType = exports.ConnectionErrorType = 'ConnectionError';
var UnknownSystemErrorType = exports.UnknownSystemErrorType = 'UnknownSystemError';
var InvalidRequestErrorType = exports.InvalidRequestErrorType = 'InvalidRequestError';
var UnsupportedPhotoErrorType = exports.UnsupportedPhotoErrorType = 'UnsupportedPhotoError';

var UnsupportedPhotoError = exports.UnsupportedPhotoError = function (_Error) {
  _inherits(UnsupportedPhotoError, _Error);

  function UnsupportedPhotoError(message) {
    _classCallCheck(this, UnsupportedPhotoError);

    // (1)
    var _this = _possibleConstructorReturn(this, (UnsupportedPhotoError.__proto__ || Object.getPrototypeOf(UnsupportedPhotoError)).call(this, message));

    _this.type = UnsupportedPhotoErrorType;
    Object.setPrototypeOf(_this, UnsupportedPhotoError.prototype);
    return _this;
  }

  return UnsupportedPhotoError;
}(Error);

var AuthenticationError = exports.AuthenticationError = function (_Error2) {
  _inherits(AuthenticationError, _Error2);

  function AuthenticationError(message) {
    _classCallCheck(this, AuthenticationError);

    // (1)
    var _this2 = _possibleConstructorReturn(this, (AuthenticationError.__proto__ || Object.getPrototypeOf(AuthenticationError)).call(this, message));

    _this2.type = AuthenticationErrorType;
    Object.setPrototypeOf(_this2, AuthenticationError.prototype);
    return _this2;
  }

  return AuthenticationError;
}(Error);

var InvalidRequestError = exports.InvalidRequestError = function (_Error3) {
  _inherits(InvalidRequestError, _Error3);

  function InvalidRequestError(message, errors) {
    _classCallCheck(this, InvalidRequestError);

    var _this3 = _possibleConstructorReturn(this, (InvalidRequestError.__proto__ || Object.getPrototypeOf(InvalidRequestError)).call(this, message));

    _this3.type = InvalidRequestErrorType;
    if (errors) {
      errors = errors.map ? errors : [errors];
    }
    _this3.errors = errors;
    Object.setPrototypeOf(_this3, InvalidRequestError.prototype);
    return _this3;
  }

  return InvalidRequestError;
}(Error);

var ConnectionError = exports.ConnectionError = function (_Error4) {
  _inherits(ConnectionError, _Error4);

  function ConnectionError(message) {
    _classCallCheck(this, ConnectionError);

    // (1)
    var _this4 = _possibleConstructorReturn(this, (ConnectionError.__proto__ || Object.getPrototypeOf(ConnectionError)).call(this, message));

    _this4.type = ConnectionErrorType;
    Object.setPrototypeOf(_this4, ConnectionError.prototype);
    return _this4;
  }

  return ConnectionError;
}(Error);

var UnknownSystemError = exports.UnknownSystemError = function (_Error5) {
  _inherits(UnknownSystemError, _Error5);

  function UnknownSystemError(message) {
    _classCallCheck(this, UnknownSystemError);

    // (1)
    var _this5 = _possibleConstructorReturn(this, (UnknownSystemError.__proto__ || Object.getPrototypeOf(UnknownSystemError)).call(this, message));

    _this5.type = UnknownSystemErrorType;
    Object.setPrototypeOf(_this5, UnknownSystemError.prototype);
    return _this5;
  }

  return UnknownSystemError;
}(Error);
